File: annotation-processor/src/main/java/com/mayankrastogi/cs474/hw2/annotations/processor/AbstractAnnotationProcessor.java
package com.mayankrastogi.cs474.hw2.annotations.processor;

import javax.annotation.processing.*;
import javax.lang.model.SourceVersion;
import javax.lang.model.element.*;
import javax.lang.model.type.DeclaredType;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Elements;
import javax.lang.model.util.Types;
import javax.tools.Diagnostic;
import java.lang.annotation.Annotation;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public abstract class AbstractAnnotationProcessor extends AbstractProcessor {

    private static final String DEBUG_OPTION_NAME = "AnnotationProcessor.debug";

    Elements elementUtils;
    Types typeUtils;
    boolean treatMandatoryWarningsAsErrors = false;

    private Messager messager;
    private boolean debug = false;

    @Override
    public synchronized void init(ProcessingEnvironment processingEnv) {
        super.init(processingEnv);

        messager = processingEnv.getMessager();
        elementUtils = processingEnv.getElementUtils();
        typeUtils = processingEnv.getTypeUtils();
        debug = processingEnv.getOptions().containsKey(DEBUG_OPTION_NAME);
    }

    @Override
    public Set<String> getSupportedAnnotationTypes() {
        var supportedAnnotationTypesAnnotation = this.getClass().getAnnotation(SupportedAnnotationTypes.class);

        Stream<String> superTypes = Stream.empty();
        if (supportedAnnotationTypesAnnotation != null)
            superTypes = Arrays.stream(supportedAnnotationTypesAnnotation.value());

        return Stream.concat(superTypes, getSupportedAnnotations().map(Class::getCanonicalName))
                .collect(Collectors.toUnmodifiableSet());
    }

    protected Stream<Class> getSupportedAnnotations() {
        return Stream.empty();
    }

    @Override
    public Set<String> getSupportedOptions() {
        return Stream.of(DEBUG_OPTION_NAME).collect(Collectors.toUnmodifiableSet());
    }

    @Override
    public SourceVersion getSupportedSourceVersion() {
        var sourceVersionAnnotation = this.getClass().getAnnotation(SupportedSourceVersion.class);
        return sourceVersionAnnotation == null ? SourceVersion.RELEASE_11 : sourceVersionAnnotation.value();
    }

    // =================================================================================================================
    // Utility functions for extracting common information from elements
    // =================================================================================================================

    boolean elementImplementsInterface(Element element, String interfaceClassName) {
        var elementType = typeUtils.erasure(element.asType());
        var iteratorType = typeUtils.erasure(elementUtils.getTypeElement(interfaceClassName).asType());

        return typeUtils.isSubtype(elementType, iteratorType);
    }

    boolean assertElementDefinesMethod(Element element, String methodName, TypeMirror returnType, List<? extends VariableElement> parameters) {
        var methodFound = element.getEnclosedElements()
                .stream()
                .filter(e -> e instanceof ExecutableElement)
                .map(e -> (ExecutableElement) e)
                .filter(e ->
                        e.getSimpleName().toString().equals(methodName) &&
                                typeUtils.isSameType(e.getReturnType(), returnType) &&
                                e.getParameters().equals(parameters))
                .count();
        if (methodFound != 1) {
            error(String.format("%s does not define the method `%s` with return type `%s` and parameters `%s`", element, methodName, returnType, parameters), element);
            return false;
        }
        return true;
    }

    Optional<? extends TypeMirror> getTypeParameterFromImplementedInterface(Element element, String interfaceClassName) {
        return getTypeParameterFromImplementedInterface(element, elementUtils.getTypeElement(interfaceClassName).asType());
    }

    Optional<? extends TypeMirror> getTypeParameterFromImplementedInterface(Element element, TypeMirror interfaceType) {
        if (element instanceof TypeElement) {
            return ((TypeElement) element).getInterfaces()
                    .stream()
                    .filter(i -> typeUtils.isSameType(typeUtils.erasure(i), typeUtils.erasure(interfaceType)))
                    .flatMap(this::getTypeParametersFromGenericElement)
                    .findFirst();
        }
        return Optional.empty();
    }

    private Stream<? extends TypeMirror> getTypeParametersFromGenericElement(TypeMirror element) {
        if (element instanceof DeclaredType) {
            return ((DeclaredType) element).getTypeArguments().stream();
        } else {
            return Stream.empty();
        }
    }

    String getAnnotationValueAsClassName(Element element, Class<? extends Annotation> annotationClass) {
        var annotationValue = getAnnotationValueAsType(element, annotationClass);
        return annotationValue != null ? annotationValue.toString() : "";
    }

    TypeMirror getAnnotationValueAsType(Element element, Class<? extends Annotation> annotationClass) {
        var annotationMirror = getAnnotationMirror(element, annotationClass);
        if (annotationMirror.isEmpty()) return null;
        return getAnnotationValueAsType(annotationMirror.get(), "value");
    }

    // Borrowed from Dave Dopson's answer on StackOverflow: https://stackoverflow.com/a/10167558/4463881
    private TypeMirror getAnnotationValueAsType(AnnotationMirror annotationMirror, String key) {
        var annotationValue = getAnnotationValue(annotationMirror, key);
        if (annotationValue.isEmpty()) {
            return null;
        }
        return (TypeMirror) annotationValue.get().getValue();
    }

    // Borrowed from tquadrat's answer on StackOverflow: https://stackoverflow.com/a/52257877/4463881
    private Optional<? extends AnnotationMirror> getAnnotationMirror(Element element, Class<? extends Annotation> annotationClass) {
        var annotationClassName = annotationClass.getName();
        return element.getAnnotationMirrors().stream()
                .filter(m -> m.getAnnotationType().toString().equals(annotationClassName))
                .findFirst();
    }

    // Borrowed from tquadrat's answer on StackOverflow: https://stackoverflow.com/a/52257877/4463881
    private Optional<? extends AnnotationValue> getAnnotationValue(AnnotationMirror annotationMirror, String name) {
        var elementValues = elementUtils.getElementValuesWithDefaults(annotationMirror);
        return elementValues.keySet().stream()
                .filter(k -> k.getSimpleName().toString().equals(name))
                .map(elementValues::get)
                .findAny();
    }

    // =================================================================================================================
    // Helpers for printing diagnostic messages during annotation processing
    // =================================================================================================================

    void error(String message) {
        sendDiagnostic(Diagnostic.Kind.ERROR, message);
    }

    void error(String message, Element element) {
        sendDiagnostic(Diagnostic.Kind.ERROR, message, element);
    }

    void debug(String message) {
        if (debug) sendDiagnostic(Diagnostic.Kind.OTHER, "[DEBUG] " + message);
    }

    void debug(String message, Element element) {
        if (debug) sendDiagnostic(Diagnostic.Kind.OTHER, "[DEBUG]" + message, element);
    }

    void note(String message) {
        sendDiagnostic(Diagnostic.Kind.NOTE, message);
    }

    void note(String message, Element element) {
        sendDiagnostic(Diagnostic.Kind.NOTE, message, element);
    }

    void mandatoryWarning(String message) {
        if (treatMandatoryWarningsAsErrors)
            error(message);
        else
            sendDiagnostic(Diagnostic.Kind.MANDATORY_WARNING, message);
    }

    void mandatoryWarning(String message, Element element) {
        if (treatMandatoryWarningsAsErrors)
            error(message, element);
        else
            sendDiagnostic(Diagnostic.Kind.MANDATORY_WARNING, message, element);
    }

    void warning(String message) {
        sendDiagnostic(Diagnostic.Kind.WARNING, message);
    }

    void warning(String message, Element element) {
        sendDiagnostic(Diagnostic.Kind.WARNING, message, element);
    }

    private void sendDiagnostic(Diagnostic.Kind kind, String message) {
        messager.printMessage(kind, message);
    }

    private void sendDiagnostic(Diagnostic.Kind kind, String message, Element element) {
        messager.printMessage(kind, message, element);
    }
}

File: annotation-processor/src/main/scala/com/mayankrastogi/cs474/hw2/annotations/processor/AbstractAnnotationProcessor.scala
package com.mayankrastogi.cs474.hw2.annotations.processor

import java.lang.annotation.Annotation
import java.util

import javax.annotation.processing._
import javax.lang.model.SourceVersion
import javax.lang.model.`type`.TypeMirror
import javax.lang.model.element._
import javax.lang.model.util.{Elements, Types}
import javax.tools.Diagnostic

import scala.jdk.CollectionConverters._

/**
 * Provides helper methods and convenience methods that can be used by child annotation processors for performing
 * common tasks and easily print logs at the required level.
 */
abstract class AbstractAnnotationProcessor extends AbstractProcessor {

  /**
   * If the annotation processor is passed this option, irrespective of its value, debug messages are also logged during
   * the annotation processing.
   */
  private val DEBUG_OPTION_NAME = "AnnotationProcessor.debug"

  private[processor] var elementUtils: Elements = _
  private[processor] var typeUtils: Types = _
  private[processor] var treatMandatoryWarningsAsErrors = false

  private var messager: Messager = _
  private var debug = false

  override def init(processingEnv: ProcessingEnvironment): Unit = {
    super.init(processingEnv)

    messager = processingEnv.getMessager
    elementUtils = processingEnv.getElementUtils
    typeUtils = processingEnv.getTypeUtils
    debug = processingEnv.getOptions.containsKey(DEBUG_OPTION_NAME)
  }

  override def getSupportedAnnotationTypes: util.Set[String] = {
    // Get the annotations specified in @SupportedAnnotationTypes, if present on the annotation processor
    val supportedAnnotationTypesAnnotation = getClass.getAnnotation(classOf[SupportedAnnotationTypes])
    val superTypes =
      if (supportedAnnotationTypesAnnotation != null)
        supportedAnnotationTypesAnnotation.value
      else
        Array.empty[String]

    // Append this with the list of annotations specified in the `getSupportedAnnotations` method
    (getSupportedAnnotations.map(_.getCanonicalName) ++ superTypes).toSet.asJava
  }

  /**
   * Specifies the list of annotations that can be processed by this annotation processor.
   *
   * This list is merged with list of class names specified in the @[[SupportedAnnotationTypes]], if present.
   *
   * @return A list of classes of annotation types.
   */
  protected def getSupportedAnnotations: List[Class[_ <: Annotation]] = List.empty

  override def getSupportedOptions: util.Set[String] = Set(DEBUG_OPTION_NAME).asJava

  override def getSupportedSourceVersion: SourceVersion = {
    // If a value is specified in @SupportedSourceVersion, use that value, else default to SourceVersion.RELEASE_11
    val sourceVersionAnnotation = getClass.getAnnotation(classOf[SupportedSourceVersion])
    if (sourceVersionAnnotation == null) SourceVersion.RELEASE_11 else sourceVersionAnnotation.value
  }

  // ===================================================================================================================
  // Utility functions for extracting common information from elements
  // ===================================================================================================================

  /**
   * Performs an operation while treating any calls to `warning()` during this operation as if a call to `error()` was
   * made, if the optional `flag` is set to `true` (default).
   *
   * @param flag      Whether warnings should be treated as errors during the operation.
   * @param operation A block of code to execute, which may call the [[warning()]] method during its execution.
   * @tparam T The return type of the `operation`.
   * @return The value returned by the `operation` upon execution.
   */
  private[processor] def withTreatMandatoryWarningsAsErrors[T](flag: Boolean = true)(operation: => T): T = {
    // Store current state
    val previousState = treatMandatoryWarningsAsErrors

    // Change the state with new value and perform the operation
    treatMandatoryWarningsAsErrors = flag
    val result = operation

    // Restore the state
    treatMandatoryWarningsAsErrors = previousState
    result
  }

  /**
   * Extracts the `value()` property as a [[String]] from the annotation of the specified type, applied on the supplied
   * `element`.
   *
   * @param element         The [[Element]] that is annotated with the annotation of `annotationClass`.
   * @param annotationClass The [[Class]] of the annotation whose `value()` will be retuned.
   * @return The value of the `value()` property as a [[String]].
   */
  private[processor] def getAnnotationValueAsClassName(element: Element, annotationClass: Class[_ <: Annotation]): String = {
    val annotationValue = getAnnotationValueAsType(element, annotationClass)

    if (annotationValue != null) annotationValue.toString else ""
  }

  /**
   * Extracts the `value()` property as a [[TypeMirror]] from the annotation of the specified type, applied on the
   * supplied `element`.
   *
   * @param element         The [[Element]] that is annotated with the annotation of `annotationClass`.
   * @param annotationClass The [[Class]] of the annotation whose `value()` will be retuned.
   * @return The [[TypeMirror]] corresponding to the value of the `value()` on the specified annotation.
   */
  private[processor] def getAnnotationValueAsType(element: Element, annotationClass: Class[_ <: Annotation]): TypeMirror = {
    val annotationMirror = getAnnotationMirror(element, annotationClass)

    if (annotationMirror.isDefined) getAnnotationValueAsType(annotationMirror.get, "value") else null
  }

  /**
   * Extracts a value as a [[TypeMirror]] from the specified [[AnnotationMirror]].
   *
   * Borrowed from Dave Dopson's answer on StackOverflow: [[https://stackoverflow.com/a/10167558/4463881]].
   *
   * @param annotationMirror The annotation object from which the the value is to be extracted.
   * @param key              The name of the property, defined in the annotation, which contains the required value.
   * @return The value stored in the given `key` in the supplied `annotationMirror` as a [[TypeMirror]].
   */
  private def getAnnotationValueAsType(annotationMirror: AnnotationMirror, key: String): TypeMirror = {
    val annotationValue = getAnnotationValue(annotationMirror, key)

    if (annotationValue.isDefined) annotationValue.get.getValue.asInstanceOf[TypeMirror] else null
  }

  /**
   * Extracts a value from the specified property `name` defined in the supplied [[AnnotationMirror]].
   *
   * @param annotationMirror The annotation object from which the the value is to be extracted.
   * @param name             The name of the property, defined in the annotation, which contains the required value.
   * @return The value stored in the given `key` in the supplied `annotationMirror`.
   */
  private def getAnnotationValue(annotationMirror: AnnotationMirror, name: String): Option[_ <: AnnotationValue] = {
    val elementValues =
      elementUtils.getElementValuesWithDefaults(annotationMirror)
        .asScala
        .toMap[ExecutableElement, AnnotationValue]

    elementValues
      .keySet
      .filter(_.getSimpleName.toString == name)
      .map(elementValues.get)
      .head
  }

  /**
   * Gets the instance of the given `annotationClass` which is applied on the specified `element`.
   *
   * Borrowed from tquadrat's answer on StackOverflow: [[https://stackoverflow.com/a/52257877/4463881]].
   *
   * @param element         The [[Element]] that is annotated with the annotation of `annotationClass`.
   * @param annotationClass The [[Class]] of the annotation which is applied on the given `element`.
   * @return The instance of the `annotationClass` applied on the specified `element`.
   */
  private def getAnnotationMirror(element: Element, annotationClass: Class[_ <: Annotation]): Option[_ <: AnnotationMirror] = {
    element
      .getAnnotationMirrors.asScala
      .find(_.getAnnotationType.toString == annotationClass.getName)
  }

  // ===================================================================================================================
  // Helpers for printing diagnostic messages during annotation processing
  // ===================================================================================================================

  private[processor] def debug(message: String, element: Element = null): Unit = {
    if (debug) sendDiagnostic(Diagnostic.Kind.OTHER, "[DEBUG]" + message, element)
  }

  private[processor] def note(message: String, element: Element = null): Unit = {
    sendDiagnostic(Diagnostic.Kind.NOTE, message, element)
  }

  private[processor] def warning(message: String, element: Element): Unit = {
    if (treatMandatoryWarningsAsErrors) error(message, element)
    else sendDiagnostic(Diagnostic.Kind.MANDATORY_WARNING, message, element)
  }

  private[processor] def error(message: String, element: Element = null): Unit = {
    sendDiagnostic(Diagnostic.Kind.ERROR, message, element)
  }

  private def sendDiagnostic(kind: Diagnostic.Kind, message: String, element: Element): Unit = {
    if (element == null)
      messager.printMessage(kind, message)
    else
      messager.printMessage(kind, message, element)
  }
}
